### PR TITLE
refactoring and typehints of ml_train

### PR DESCRIPTION
**Pull request by Vikka**
Update of code and documentation to be both less and more restrictive about what can be set to `backend.ml.handle_data.select_criteria` and what `backend.ml.management.commands.ml_train.fetch_data` returns.

### CODE REVIEW

**Comment by Vikka** [comment]
```
tournesol/entities/__init__.py:1:0: R0801: Similar lines in 2 files
==tournesol.models.entity_score:23
==tournesol.models.ratings:51
    )
    criteria = models.TextField(
        max_length=32,
        help_text="Name of the criteria",
        db_index=True,
    )
    score = models.FloatField(
        default=0,
        blank=False, (duplicate-code)
tournesol/entities/__init__.py:1:0: R0801: Similar lines in 2 files
==tournesol.models.comparisons:120
==tournesol.models.ratings:49
        related_name="criteria_scores",
        on_delete=models.CASCADE,
    )
    criteria = models.TextField(
        max_length=32,
        help_text="Name of the criteria",
        db_index=True,
    )
    score = models.FloatField( (duplicate-code)
tournesol/entities/__init__.py:1:0: R0801: Similar lines in 2 files
==tournesol.models.comparisons:122
==tournesol.models.entity_score:23
    )
    criteria = models.TextField(
        max_length=32,
        help_text="Name of the criteria",
        db_index=True,
    )
    score = models.FloatField( (duplicate-code)
tournesol/entities/__init__.py:1:0: R0801: Similar lines in 2 files
==tournesol.views.polls:89
==tournesol.views.video:173
            queryset = queryset.filter(total_score__isnull=False)
        else:
            queryset = queryset.filter(
                rating_n_contributors__gte=settings.RECOMMENDATIONS_MIN_CONTRIBUTORS
            ).filter(total_score__gt=0) (duplicate-code)

-----------------------------------
Your code has been rated at 8.69/10
```

I didn't code this, any info about this issue?

**Comment by GresilleSiffle** [comment]
We can see in the CI that the return code of `pylint` is 0, so everything is fine.

It's `isort` raising 3 errors. You can run the content of `scripts/ci/lint.sh` in local so see what's going on.

You can also run the content `scripts/ci/lint-fix.sh` to automatically fix these errors : )

**Comment by Vikka** [comment]
Oh :
```text
ERROR: /home/runner/work/tournesol/tournesol/backend/ml/handle_data.py Imports are incorrectly sorted and/or formatted.
ERROR: /home/runner/work/tournesol/tournesol/backend/ml/core.py Imports are incorrectly sorted and/or formatted.
ERROR: /home/runner/work/tournesol/tournesol/backend/ml/management/commands/ml_train.py Imports are incorrectly sorted and/or formatted.
```
Thanks

File: backend/ml/handle_data.py [hunk 1]
```diff
@@ -23,32 +23,35 @@
...
-def shape_data(l_ratings):
+def shape_data(l_ratings: Iterable[Tuple[int, int, int, str, int, int]]):
     """Shapes data for distribute_data()/distribute_data_from_save()
 
     l_ratings : list of not None ratings ([0,100]) for one criteria, all users
 
     Returns : one array with 4 columns : userID, vID1, vID2, rating ([-1,1])
     """
-    l_clear = [rating[:3] + [rescale_rating(rating[4])] for rating in l_ratings]
-    return np.asarray(l_clear)
```
Comment (Vikka) [inline backend/ml/handle_data.py:51]: https://stackoverflow.com/questions/14415741/what-is-the-difference-between-numpys-array-and-asarray-functions#comment36996511_14415801
Here, l_clear is a list, so a copy must be made.

If my work isn't conclusive, we still should replace this to a np.array call

**Comment by Vikka** [comment]
`ERROR: /home/runner/work/tournesol/tournesol/backend/ml/tests/ml_tests.py Imports are incorrectly sorted and/or formatted.`

Ok, then... 
```python
import numpy as np
import torch

from ml.core import _set_licchavi, _train_predict, ml_run
from ml.data_utility import (
    expand_dic,
    expand_tens,
    get_all_vids,
    get_mask,
    rescale_rating,
    reverse_idxs,
    sort_by_first,
)
from ml.dev.fake_data import generate_data
from ml.handle_data import distribute_data, select_criteria, shape_data
from ml.licchavi import get_model, get_s, Licchavi
from ml.losses import (
    _approx_bbt_loss,
    _bbt_loss,
    get_s_loss,
    model_norm,
    models_dist,
)
from ml.metrics import (
    _random_signs,
    check_equilibrium_glob,
    check_equilibrium_loc,
    extract_grad,
    get_uncertainty_glob,
    get_uncertainty_loc,
    scalar_product,
)
```
🤔

**Comment by GresilleSiffle** [comment]
Hélo @Vikka hélo everyone

We are heavily refactoring the ML app in Django, this pull request might become obsolete.

I propose to wait for the ML refactor is done then rebase this branch on `main` if it's still relevant.

**Comment by Vikka** [comment]
@GresilleSiffle Sur, just ping me when it's done :)

**Comment by GresilleSiffle** [comment]
Hello @Vikka 

We have finished our major refactor of the ML module. More changes will happen but I think you can freely checkout the new code on main now : )

**Comment by Vikka** [comment]
Thanks, I'll check that !

**Comment by lfaucon** [comment]
This PR has been a draft for a long time and now has conflicts. It is not a priority for us to work on it to finish the contribution. Please re-open the PR if you want to finish it @Vikka

**Pull request closed without merging**